hermes/crates/hermes-ipfs/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+//! Hermes IPFS
+//!
+//! Provides support for storage, and `PubSub` functionality.
+
+use std::{
+    ops::{Deref, DerefMut},
+    str::FromStr,
+};
+
+/// IPFS Content Identifier.
+pub use libipld::{self as ipld, Cid};
+/// Peer Info type.
+pub use rust_ipfs::p2p::PeerInfo;
+/// Enum for specifying paths in IPFS.
+pub use rust_ipfs::path::IpfsPath;
+/// Multiaddr type.
+pub use rust_ipfs::Multiaddr;
+/// Peer ID type.
+pub use rust_ipfs::PeerId;
+/// Builder type for IPFS Node configuration.
+pub use rust_ipfs::UninitializedIpfsNoop as IpfsBuilder;
+use rust_ipfs::{unixfs::AddOpt, Ipfs};
+
+/// Hermes IPFS Node
+///
+/// Provides the functionality of the inner `IPFS` by de-referencing.
+#[derive(Clone, Debug)]
+pub struct Node(Ipfs);
+
+impl From<Ipfs> for Node {
+    fn from(value: Ipfs) -> Self {
+        Self(value)
+    }
+}
+
+impl Deref for Node {
+    type Target = Ipfs;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for Node {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// Hermes IPFS
+#[allow(dead_code)]
+pub struct HermesIpfs {
+    /// IPFS node
+    pub node: Node,
+}
+
+impl HermesIpfs {
+    /// Start a new node.
+    ///
+    /// ## Returns
+    ///
+    /// * `HermesIpfs`
+    ///
+    /// ## Errors
+    ///
+    /// Returns an error if the IPFS daemon fails to start.
+    pub async fn start() -> anyhow::Result<Self> {
+        // TODO(saibatizoku):
+        let node: Ipfs = IpfsBuilder::new()
+            .with_default()
+            .set_default_listener()
+            //.with_mdns()
+            //.with_upnp()
+            .start()
+            .await?;
+
+        Ok(HermesIpfs { node: Node(node) })
+    }
+
+    /// Add a file to IPFS.
+    ///
+    /// ## Parameters
+    ///
+    /// * `file_path` The `file_path` can be specified as a type that converts into
+    ///   `std::path::PathBuf`.
+    ///
+    /// ## Returns
+    ///
+    /// * A result with `IpfsPath`
+    ///
+    /// ## Errors
+    ///
+    /// Returns an error if the file fails to upload.
+    pub async fn add_ipfs_file(&self, ipfs_file: AddIpfsFile) -> anyhow::Result<IpfsPath> {
+        let ipfs_path = self.node.add_unixfs(ipfs_file).await?;
+        Ok(ipfs_path)
+    }
+
+    /// Get a file from IPFS
+    ///
+    /// ## Parameters
+    ///
+    /// * `ipfs_path` - `GetIpfsFile(String)` Path used to get the file from IPFS.
+    ///
+    /// ## Returns
+    ///
+    /// * `A result with Vec<u8>`.
+    ///
+    /// ## Errors
+    ///
+    /// Returns an error if the file fails to download.
+    pub async fn get_ipfs_file(&self, ipfs_path: GetIpfsFile) -> anyhow::Result<Vec<u8>> {
+        let stream_bytes = self.node.cat_unixfs(ipfs_path).await?;
```

Comment:
Is it possible to return here just a bytes stream ?
For the case when a file is really huge I think this approach could lead to performance issues.

Reply:
This method should just return the resulting bytes from IPFS, under the hood, it is implemented as an async stream. However, if the use case for handling the stream is required, the user can invoke `self.node.cat_unixfs` directly, and deal with it as needed.

Reply:
I think this should stay as it is, and if we find issues we can revise.